modules/navier_stokes/src/fvkernels/PINSFVMomentumDiffusion.C

```diff
@@ -87,33 +77,50 @@ PINSFVMomentumDiffusion::computeQpResidual()
               true);
 
   // Compute face superficial velocity gradient
-  auto dudn = gradUDotNormal();
+  auto dudn =
+      _var.gradient(Moose::FV::makeCDFace(*_face_info, faceArgSubdomains())) * _face_info->normal();
+
+  if (_computing_rc_data)
+  {
+    if (_face_type == FaceInfo::VarFaceNeighbors::ELEM ||
+        _face_type == FaceInfo::VarFaceNeighbors::BOTH)
+    {
+      const auto dof_number = _face_info->elem().dof_number(_sys.number(), _var.number(), 0);
+      // A gradient is a linear combination of degrees of freedom so it's safe to straight-up index
+      // into the derivatives vector at the dof we care about
+      _ae = dudn.derivatives()[dof_number];
+      _ae *= -mu_eps_face;
+    }
+    if (_face_type == FaceInfo::VarFaceNeighbors::NEIGHBOR ||
+        _face_type == FaceInfo::VarFaceNeighbors::BOTH)
+    {
+      const auto dof_number = _face_info->neighbor().dof_number(_sys.number(), _var.number(), 0);
+      _an = dudn.derivatives()[dof_number];
+      _an *= mu_eps_face;
+    }
+  }
 
   // First term of residual
   ADReal residual = mu_eps_face * dudn;
 
   if (_smooth_porosity)
   {
     // Get the face porosity gradient separately
-    const auto & grad_eps_face = MetaPhysicL::raw_value(_eps_var->adGradSln(*_face_info));
-
-    ADRealVectorValue term_elem = mu_elem / eps_elem / eps_elem * grad_eps_face;
-    ADRealVectorValue term_neighbor = mu_neighbor / eps_neighbor / eps_neighbor * grad_eps_face;
+    const auto & grad_eps_face =
+        MetaPhysicL::raw_value(_eps.gradient(Moose::FV::makeCDFace(*_face_info)));
 
-    const auto vel_elem = (*_vel)(elem_face);
-    const auto vel_neighbor = (*_vel)(neighbor_face);
-
-    for (int i = 0; i < LIBMESH_DIM; i++)
-    {
-      term_elem(i) *= vel_elem(i);
-      term_neighbor(i) *= vel_neighbor(i);
-    }
+    const auto coeff_elem = mu_elem / eps_elem / eps_elem * _var(elem_face);
+    const auto coeff_neighbor = mu_neighbor / eps_neighbor / eps_neighbor * _var(neighbor_face);
 
     // Interpolate to get the face value
-    ADRealVectorValue term_face;
-    interpolate(
-        Moose::FV::InterpMethod::Average, term_face, term_elem, term_neighbor, *_face_info, true);
-    residual -= term_face * _normal;
+    ADReal coeff_face;
+    interpolate(Moose::FV::InterpMethod::Average,
+                coeff_face,
+                coeff_elem,
+                coeff_neighbor,
+                *_face_info,
+                true);
+    residual -= coeff_face * grad_eps_face * _normal;
```

Comment:
@GiudGiud I changed this code up and I now get second order convergence for all PINSFV variables for a continuously varying porosity MMS case. I don't think the loop over `LIBMESH_DIM` was quite right. As long as we are doing the "Laplace" form for stress, then this calculation should only involve the current velocity component. This change required regolding one test (the `porosity_jump.smooth_jump` test), and I think it looks a little bit better now (although really not a huge change). Here is the pressure along the flow direction where the orange line is with the new code, and the blue line was the previous code.

<img width="1312" alt="Screen Shot 2021-11-22 at 8 50 27 PM" src="https://user-images.githubusercontent.com/10248304/142973428-7c41342c-497d-40e9-ad95-79ed0191d758.png">

It's not super easy to see but before the discontinuity the old-way pressure drops a little below the asymptotic entry pressure and then after the discontinuity the old-way pressure drops a little below the asymptotic exit pressure. Below I've zoomed in on the region between the discontinuity and exit. (Even zoomed in it's still not easy to see; you'll probably have to click on the figure to see it)

<img width="1316" alt="Screen Shot 2021-11-22 at 8 50 14 PM" src="https://user-images.githubusercontent.com/10248304/142973391-b8598fc0-73af-4598-bf12-559cec1355bf.png">

Reply:
Interesting. I just tried the derivation again and got the same result as I got last time :(

Reply:
So what i did was take the X-component of the derivation of grad(\vec{v}/eps) \cdot n (for the IBP or div(grad(\vec{v}/eps)).
This looks more like grad(v_x/eps) \cdot n. This is consistent with div(grad(v_x/eps))

I cant remember how I got to my form. I think I started from Stokes s stress equation.

Reply:
So the below leverages the identities from [here](https://en.wikipedia.org/wiki/Vector_calculus_identities#First_derivative_identities)

![pinsfv-diff-derivation](https://user-images.githubusercontent.com/10248304/143073292-976d669e-3c9e-4070-99ad-366918797d14.jpg)

Reply:
Agree with line 1 and 2.
Where we differ is line 3 where you do n.A and I did A.n. I m more used to the latter for divergence theorem

Reply:
When A is a vector field, writing as n.A vs. A.n makes no difference. However, obviously the ordering is very important when A is a tensor (matrix here). See my derivation below for divergence of a tensor. I'm sure we can agree that the divergence operator must be applied on the left hand side. Consequently, the divergence operator ends up operating on the column vectors of the matrix. So the normal vector should also end up operating on the column vectors of the matrix.

![div-with-a-tensor](https://user-images.githubusercontent.com/10248304/143080992-1faf96ce-7af2-41e1-9606-eec175c76ea8.jpg)

Reply:
well both definitions exist for div() and nabla dot.
If you want to end up with a column vector you need to apply it on the right side? What you wrote should be a row vector

Reply:
I think convention for nabla dot is left side though

Reply:
ok so divergence theorem is indeed with n . A. Tensor math is hard

Reply:
> If you want to end up with a column vector you need to apply it on the right side? What you wrote should be a row vector

You are right that post matrix-vector product, I should have been writing a 1x2 matrix (vector) instead of a 2x1 matrix (vector), so the final result should be (n * c1, n * c2) instead of the transpose of that.

Reply:
> ok so divergence theorem is indeed with n . A. Tensor math is hard

So we are in agreement now?

Reply:
Yes I think so

Reply:
looking at the divergence theorem for tensors also brings up the other formula, with n on the right. 

What it boils down to is that in the Navier Stokes equation it's 
d/dxj ( mu (dui/dxj + duj/dxi) )
But that is actually the classic definition of divergence, the one that has the following divergence theorem:

int_vol dj T_ij dV = int_surface Tij nj dS
so for each component i, the vector is Tij nj, or a sum over the row

so we dont agree anymore haha
but also there s a high likelihood I m wrong and a total certainty I have to work on something else right now. 
I promise to look again before i review your PR

Reply:
[This](https://pkel015.connect.amazon.auckland.ac.nz/SolidMechanicsBooks/Part_III/Chapter_1_Vectors_Tensors/Vectors_Tensors_14_Tensor_Calculus.pdf) is a really nice reference. Ok, as stated in that reference and as also stated in [Wikipedia's article](https://en.wikipedia.org/wiki/Divergence#Tensor_field), div(T^t) = \nabla \cdot T. And you are right that in the literature I am also seeing:

\int div(T) dV = \int Tij nj dS

so right multiplication by the normal vector. However, if you agreed with me on the first few steps of my derivation, then we must transpose my matrix before right multiplying by the normal vector, since the Navier-Stokes equations are written as \nabla \cdot \nabla u, not as div(\nabla u). So given this the result would be the same, in favor of my result 😄 .

Note that if we are indeed working with the [incompressible form of the stress term](https://en.wikipedia.org/wiki/Navier%E2%80%93Stokes_equations#Incompressible_flow), and we assume a constant mu, then for this:

\nabla \cdot \mu (\nabla u + (\nabla u)^t)

the transpose part is the part that vanishes to 0 because of the incompressibility constraint, such that we are left with

\nabla \cdot \mu \nabla u

Reply:
Yes it boils down to what is the Navier Stokes equation term for the stress term \nabla \cdot \mu \nabla u or div(\mu \nabla u). It s the same for most most people, not for us.

You agree this term is written : sum_j d/dxj ( mu (dui/dxj ) e_i ) (ignore the transpose part) ?
For me this is correct because it reduces to the laplacian as expected, for each component. 

If so, then that's the divergence, not the nabla. Nabla would be:
sum_i d/dxi ( mu (dui/dxj ) e_j )
or at least according to the wiki page. 

Then maybe there's a transpose operation to take on that but I dont see why

Reply:
umm my expression for nabla doesnt really work does it. It's the wrong equation, we dont fall back on our feet from the symmetry of the stress tensor. Well I ll leave it at that and see what you think

Reply:
Let's forget about the tensor math for a second. Can we agree that with the incompressible constraint applied, then the component-wise momentum equation for PINSFV looks like so?
```
div(vel*rho*u/porosity) - div(mu*grad(u/porosity)) + porosity*grad(p).dot(e_i)
```
This is indeed what you have in your own 1D MMS test (and should have had in your 2D MMS tests). In that case, I think it's pretty clear that my result, with each momentum component equation having a grad_eps term from the viscous term, is the correct one.

Reply:
If we can't agree on the component-wise form, then we should have never have been just using `FVDiffusion` for our INSFV simulations

Reply:
That's an equivalent formulation of the problem. If we just write NS directly with u instead of vec_u then we are good to go with your form. 1D absolutely reduces to this.

I think what I wrote had a grad_eps term for each component as well. 

I dont see why we cant use FVDiffusion if we start from the vector NS equation.

Reply:
What you had before didn't match either left multiplication or right multiplication of the tensor by the normal. Depending on how you multiply the tensor by the normal your options for the result are for a given momentum component:

mu/eps^2 u_i * n * grad_eps (what I am almost 100% is the correct way)

or

mu/eps^2 * grad_eps_i * n * u (which simply does not match any term that would fall out of n * mu * grad(u_i / eps) which corresponds to n * mu * grad(u_free_i) ... which corresponds to `FVDiffusion` ... which we've used for all of our free flow simulation thus far because *it is* the viscous Navier-Stokes term with the incompressibility constraint applied, and the viscous term for free-flow and porous flow should 100% be the same based on all the equations I've seen thus far)

Reply:
grad(u_i / eps) is not the same as (grad(u / eps) )_i
I think what I have is mu/eps^2 * (grad_eps u) * n

Reply:
I ll double check that last statement in like an hour.

Reply:
> grad(u_i / eps) is not the same as (grad(u / eps) )_i

Most certainly not. One is a vector and one is a scalar. When you do ever get a (grad(u/eps))_i anywhere?

Reply:
The current state of the kernel is doing an inner product of grad_eps and the velocity when you should be doing an outer product

Reply:
An outer product between vectors should yield a tensor, not a vector

Reply:
If you can prove to me that an alternative formulation yields on a scalar component basis mu \nabla u_i_free like my formulation does then I'll be ready to acknowledge that it is equally valid
![free-flow-laplacian-proof](https://user-images.githubusercontent.com/10248304/144298148-1264be3e-818d-41fb-ac67-2e31a080eccf.jpg)

Reply:
I did that math component-wise not using vector math, but to get to it you just need to do the right multiply by n (instead of left) in the divergence theorem iirc.
Ok lemme see about FVDiffusion first.
@grmnptr was saying in our meeting today that it should be u_D not u in the diffusion term anyway, so we may be on the verge of ditching this term entirely

Reply:
ok so for FVDiffusion:
div grad (u_vec) = div (dui dxj) = sum_i sum_j  d2ui d2xj ei 
so one FVDiffusion per component

For the form of the equation (u or u_D) inside the gradient:
in Pronghorn we have u not u_D inside the gradient. You see it in the weak form. That way there is no need for a grad_eps term. 
since we are working with u_D we need the grad_eps term. So imo we are consistent with Pgh FE right now.

Reply:
@lindsayad we should just run a porosity jump channel that is bent 30 degrees. Whichever formulation is right will give the same results there as in the straight direction

Reply:
If you are working using Carlo's first paper he has 1 typo and an invalid mathematical operation in there. I read the books he referenced and some of them actually used his formulation but they didn't consider jumps in the porosity. They moved away from that formulation about 2 years ago. The one they are using right now is in this:
https://www.sciencedirect.com/science/article/pii/S0029549321003745
However, a while back I made a comparison between the two solvers with a 2D MSFR case and they qualitatively gave more or less the same results (besides pressure of course). I wanted to summon my comparison case today but looks like they switched back to a free slack version so I can only check my messages till last January :P.

Reply:
> so one FVDiffusion per component

yep

> in Pronghorn we have u not u_D inside the gradient. You see it in the weak form. That way there is no need for a grad_eps term.

So the nonlinear variable is actually the interstitial velocity in Pronghorn FE?

> since we are working with u_D we need the grad_eps term.

Totally agree.

> So imo we are consistent with Pgh FE right now.

What do you mean by this? You've written out what the current kernel currently does and its equivalent to normal * mu * \nabla u_i for each velocity component? Because it sounds like you're telling me the Pronghorn kernel is equivalent to the free flow viscous term with the incompressibility constraint applied ... which I totally agree with

Reply:
(with normal replaced by -grad_test for a finite element formulation)

Reply:
(I've focused on the first term on the right hand side of equation 2.5 of the new paper for now. We should add the other terms generally speaking since we don't always want to assume incompressible ... )

Reply:
So there is a typo in eq (4) in the old paper. It should be div(/mu /eps /nabla u) and the invalid mathematical operation that they did in their old solver is this: div(/mu grad(u_D)) so there is a typo in (7) too. So they pulled the porosity inside the gradient. I remember seeing the difference in the paper and in the implementation and asking them which one was correct. They said the implementation was correct and the paper had the typos. This only causes differences at the interfaces.

Reply:
Well in equation 4 reproduced above 'u' represents the interstitial velocity, not the superficial velocity, so what they have written in 4 would be equivalent to \nabla \cdot \mu \nabla (\frac{u_D}{\epsilon}) *not* \nabla \cdot \mu \nabla u_D. However, that is still different from the new paper

Reply:
And I'm realizing that we never should have been neglecting the transpose part of \div(\epsilon \mu (\grad u + (\grad u)^T)) , even for incompressible flows, because you no longer get the opportunity to commute partial derivatives and hence can't apply the incompressibility constraint

Reply:
And in the case that we are no longer dropping the transpose part of that, the argument about left or right multiplication no longer matters

Reply:
The transposed term usually falls off from the mass conservation equation

Reply:
So u_D = u * gamma which can make div( mu gamma (grad u)) -> div(mu grad(gamma u) -> div(mu grad(u_D)) through a series of mathematical inaccuracies.

Reply:
> The transposed term usually falls off from the mass conservation equation

See my text:

> you no longer get the opportunity to commute partial derivatives and hence can't apply the incompressibility constraint

By the incompressibility constraint I mean the mass conservation equation with the assumption of incompressibility applied to it. For free-flows it's easy to show that application of \div(u) = 0 to the viscous term results in the removal of the transpose term after commuting partial derivatives (you can only do this if you have constant viscosity). 

For porous flows you would similarly only be able to apply div(\eps u) = 0 and remove the transpose term *if* both viscosity and porosity are constant.

Reply:
> So u_D = u * gamma which can make div( mu gamma (grad u)) -> div(mu grad(gamma u) -> div(mu grad(u_D)) through a series of mathematical inaccuracies.

I follow what you're writing, but I don't see div(mu grad(u_D)) anywhere in the 2015 text

Reply:
Just to make sure we're talking about the same paper, I'm referring to this one: https://www.sciencedirect.com/science/article/pii/S0029549315003829

Reply:
Is there a way to refresh this conversation without refreshing the page? It is annoying to wait for it to load and then click my way through this. I worked with Carlo (the author of the 2015 paper and the lead developer of GeN-Foam) on GeN-Foam for a few years. I remember seeing a difference between the paper div(mu grad(u)) and the implementation div(mu grad(U_d)). I asked him which one was correct and he said the implementation was correct and his paper has typos in it, that is the reason why you don't see my version in there.

Reply:
@grmnptr go to your email and click the link there. 
We really should be discussing on issues over PRs. PRs can actually clog up and die at too many comments, it's happened on a code review for openmc. 

We cant do the math inaccuracies, we re trying to understand what we really should be seeing. But long term maybe we should just do that and use a baffle.
If the equation is really div(mu grad(U_d)) (without any illegal manipulation) then that would mean we don't need to do any extra work and can just ditch this grad 1/eps term. 
if the equation is really div( mu gamma (grad u)) then we just need to add a porosity factor on the face of what we have right now. This is not very hard, but it is a real change, we ll have to re-gold. And it actually agrees with Pronghorn FE now that I look again. i m going to try that real quick on SANA see if that s what we're missing

sorry for half reading your message Alex. Yeah we can no longer switch div and grad and get incompressibility to simplify here. But technically we could not do that for WCNS either. I think this is just something that is being done regardless, in the same order as neglecting kinetic energy.

Reply:
> we could not do that for WCNS either

you're spot-on

> So looks like we re up for adding a div(T(gradu)) term

Yes, I really think we should. We can always check to see how much it modifies the results. Honestly it may not modify it significantly, but I'd still like to honor the governing equations (unless the results were really close and we doubled our run-time 😆 )

Reply:
> if the equation is really div( mu gamma (grad u)) then we just need to add a porosity factor on the face of what we have right now. This is not very hard, but it is a real change, we ll have to re-gold. And it actually agrees with Pronghorn FE now that I look again.

Based on this discussion and our slack discussion including @grmnptr it sounds like this is exactly what we need to do. Anyway I think this'll be my last comment on this thread. We can have more discussion on slack and/or in an issue like @GiudGiud suggested

Reply:
Ok I ll create an issue